example/advection/advection_driver.cpp

```diff
@@ -118,27 +119,68 @@ TaskCollection AdvectionDriver::MakeTaskCollection(BlockList_t &blocks, const in
     auto update = tl.AddTask(avg_data, UpdateIndependentData<MeshData<Real>>, mc0.get(),
                              mdudt.get(), beta * dt, mc1.get());
   }
+
+  const auto &buffer_send_pack =
+      blocks[0]->packages["advection_package"]->Param<bool>("buffer_send_pack");
+  if (buffer_send_pack) {
+    TaskRegion &tr = tc.AddRegion(num_partitions);
+    for (int i = 0; i < num_partitions; i++) {
+      auto &mc1 = pmesh->mesh_data.GetOrAdd(stage_name[stage], i);
+      tr[i].AddTask(none, parthenon::cell_centered_bvars::SendBoundaryBuffers, mc1);
+    }
+  } else {
+    TaskRegion &tr = tc.AddRegion(num_task_lists_executed_independently);
+    for (int i = 0; i < blocks.size(); i++) {
```

Comment:
I think that loop needs to go over all blocks, as otherwise some blocks may not get linked to a task.
In principle, the number of task lists and the number of block are completely independent from each other.
In our current example, we use `blocks.size()` independent task lists, but we could also have fewer.
In the case that there are fewer task lists, we'd still need to assign either all blocks  (to fewer task lists) or only assign a specific subset of blocks to to a task.
Here, all blocks should execute that task, so we need to loop over all blocks.

Reply:
> Here, all blocks should execute that task, so we need to loop over all blocks.

Right that's why I said it's the same in this case.

> I think that loop needs to go over all blocks, as otherwise some blocks may not get linked to a task.

Since you're calling `AddTask` the loop should go over all tasks, which is tied to all partitions/meshdata right? Or am I missing something?

Reply:
This all depends on the actual use case.
In the current use case "a collection of tasks should be executed on all blocks" I think it's important to keep the loop as is (to emphasize that this happens for all blocks).
Here are two (theoretical) example where I could imagine this would be different
a) we only want N task lists because we have N host threads that execute the task loop but the task is still executed for all blocks
then the loop may look like 
```c++
TaskRegion &tr = tc.AddRegion(N);
for (int i = 0; i < blocks.size(); i++) {
  tr[i%N].AddTask(none, &MeshBlockData<Real>::SendBoundaryBuffers, sc1.get());
```

or b) we only want to execute task on a subset of blocks that contain certain properties (but all tasks are independent and could be executed concurrently
then the loop may look like
```c++
const int M = getNumberOfBlocksWith(PropertyX);
TaskRegion &tr = tc.AddRegion(M);
for (int i = 0; i < blocks.size(); i++) {
  if (block.has(PropertyX) {
    tr[i].AddTask(none, &MeshBlockData<Real>::SendBoundaryBuffers, sc1.get());
  }
```

In both cases would loop over all blocks (or in the second case just have a loop over the blocks with that property.
Does this make sense? Maybe I'm also missing sth.

Reply:
Hmm.... I hadn't thought of case (b). In case (a), I would argue the better way to design that would be to partition the grid into N `MeshData` objects, and add the the tasks per-meshdata.

In both cases, what you prescribe works if you don't have any dependencies. If have you have dependencies, then you would need to keep track of dependencies *per meshblock* as opposed to *per call to `AddTask`*, which would get cumbersome quickly. For that reason, even case (b), you might want to restructure the tasks over `MeshData` collections instead of over `MeshBlockData`s.

Reply:
> Hmm.... I hadn't thought of case (b). In case (a), I would argue the better way to design that would be to partition the grid into N `MeshData` objects, and add the the tasks per-meshdata.

I agree that there are better approaches, but this an example on how the task infrastructure could be used.
For this reason (demonstration purpose), there's a more verbose comment in the code further up
```c++

  // Number of task lists that can be executed indepenently and thus *may*
  // be executed in parallel and asynchronous.
  // Being extra verbose here in this example to highlight that this is not
  // required to be 1 or blocks.size() but could also only apply to a subset of blocks.
  auto num_task_lists_executed_independently = blocks.size();
  TaskRegion &async_region1 = tc.AddRegion(num_task_lists_executed_independently);

  for (int i = 0; i < blocks.size(); i++) {
```

> In both cases, what you prescribe works if you don't have any dependencies. If have you have dependencies, then you would need to keep track of dependencies _per meshblock_ as opposed to _per call to `AddTask`_, which would get cumbersome quickly. For that reason, even case (b), you might want to restructure the tasks over `MeshData` collections instead of over `MeshBlockData`s.

These would end up in different `TaskRegions` that we guarantee to be executed sequentially. This is also way the current example implemented here with the two types of interfaces (MeshData and MeshBlockData) in different `TaskRegions` works (as it implicitly guarantees that dependency between regions).

Reply:
> These would end up in different TaskRegions that we guarantee to be executed sequentially. 

No not necessarily. Not if you're doing things like
```C++
tl[i%N].AddTask()
```

That's why I'm emphasizing you should always do the loop over tasks, not over blocks. In the case of example (b) this could be fixed as
```C++
const int M = getNumberOfBlocksWith(PropertyX);
TaskRegion &tr = tc.AddRegion(M);
int j = 0;
for (int i = 0; i < blocks.size(); i++) {
  if (block.has(PropertyX) {
    tr[j++].AddTask(none, &MeshBlockData<Real>::SendBoundaryBuffers, sc1.get());
  }
```

Reply:
Now I'm confused and am not sure we're talking about the same thing.
I understood your initial comment as "that `for` loop should not go over all blocks". That was the baseline for all my replies (-> "but we need to handle all blocks so the loop should [generally] include all blocks").

Could you maybe rephrase your initial comment and/or provide additional context/explanation? Or we can talk about it on Wednesday during the Profiling call.

Reply:
I guess my comment is that the for loop should (in general) loop over task lists, not blocks. In the case of this particular loop, and indeed all loops where you're doing all tasks on all blocks, they're the same.

But if you have a different number of task lists than you do `MeshBlockData`/`MeshData` objects, you're going to run into trouble in weird and subtle ways.

It's not super important. We don't need to discuss it over the call.